homeassistant/components/adaptive_lighting/switch.py

```diff
@@ -0,0 +1,646 @@
+        'off' → 'on' event and then check whether the light is still 'on' or
+        if the brightness is still decreasing. Only if it is the case we
+        adjust the lights.
+        """
+        ts_on_to_off, id_on_to_off = self._on_to_off_event.get(entity_id, (0, None))
+        id_turn_off, transition = self._turn_off_service_event.get(
+            entity_id, (None, None)
+        )
+        if (
+            id_on_to_off is not None
+            and id_turn_off is not None
+            and id_on_to_off == id_turn_off
+        ):
+            # State change 'off' → 'on' and 'light.turn_off(..., transition=...)' are
+            # from the same event, so wait at least the 'turn_off' transition time.
+            delay = transition
+        elif ts_on_to_off == 0:
+            # No state change has been registered before.
+            return False
+        else:
+            # State change 'off' → 'on' happened but **not** because a
+            # 'light.turn_off' event that is called with 'transition'.
+            delay = TURNING_OFF_DELAY
+
+        delta_time = now_ts - ts_on_to_off
+        if delta_time < delay:
+            delay -= delta_time  # already been delta_time since the event
+            brightness_going_down = True  # this might not be the case
+            _LOGGER.debug(
+                "%s: Waiting with adjusting '%s' for %s.", self._name, entity_id, delay
+            )
+            current_state = self.hass.states.get(entity_id)
+            _LOGGER.debug(
+                "%s: '%s' state before sleep is '%s'",
+                self._name,
+                entity_id,
+                current_state,
+            )
+            for _ in range(3):
+                # It can happen that the actual transition time is longer than
+                # the specified time in the 'turn_off' service, so we check
+                # whether the brightness is still going down, if so, we wait a
+                # little longer.
+                await asyncio.sleep(delay)
+                await self.hass.services.async_call(
+                    HA_DOMAIN,
+                    SERVICE_UPDATE_ENTITY,
+                    {ATTR_ENTITY_ID: entity_id},
+                    blocking=True,
+                )
+                old_state = current_state
+                current_state = self.hass.states.get(entity_id)
+                old_brightness = old_state.attributes.get(ATTR_BRIGHTNESS, 0)
+                current_brightness = current_state.attributes.get(ATTR_BRIGHTNESS, 0)
+                brightness_going_down = old_brightness > current_brightness
+                _LOGGER.debug(
+                    "%s: '%s' state after sleep is '%s'",
+                    self._name,
+                    entity_id,
+                    current_state,
+                )
+                if not brightness_going_down:
+                    break
+                delay = TURNING_OFF_DELAY
+
+            if transition is not None:
```

Comment:
@frenck, I think I have solved the problem you brought up. 

I decided to just ignore **any** 'off' -> 'on' events during the turn_off transition.

This is because in none of my attempts I can properly detect whether a light is turning off with `transition=...`.

For example, I tried:
1. wait for the transition time
2. update the light state
3. check whether the light is off or whether the brightness has decreased since the previous state access
4. if brightness has decreased wait for 5 seconds
5. repeat step 3. and 4. (up to 3 times)

I've noticed that even when calling `homeassistant.update_entity` while the light is transitioning to off, two consecutive calls (with a pause) might result in the same brightness.

Reply:
Just a comment here for myself. The sleeping should be canceled whenever a explicit "light.turn_on" event happens.￼